[SPARK-47976][Python] Add Python UDTF DataFrame API to return result of 'analyze' method as output table

dtenedor opened the pull request:
### What changes were proposed in this pull request?

This PR adds a Python UDTF DataFrame API to return the result of calling the 'analyze' method as the output table, instead of the 'eval' and 'terminate' methods.

For example:

```
val udtfRunAnalyzeMethod: UserDefinedPythonTableFunction =
  udtf.copy(returnResultOfAnalyzeMethod = true)
val df = udtfRunAnalyzeMethod(spark, lit("abc"))
  .select("schema")
  .collect()
```

This should return the 'analyze' method result schema, with something like:

```
{
  "schema": {
    "fields": [
      {
        "metadata": {},
        "name": "partition_col",
        "nullable": true,
        "type": "integer"
      },
      {
        "metadata": {},
        "name": "count",
        "nullable": true,
        "type": "integer"
      },
      {
        "metadata": {},
        "name": "last",
        "nullable": true,
        "type": "integer"
      }
    ],
    "type": "struct"
  },
  "withSinglePartition": "False",
  "partitionBy": [
    {
      "name": "partition_col"
    }
  ],
  "orderBy": [
    {
      "name": "input",
      "ascending": "True",
      "overrideNullsFirst": "None"
    }
  ],
  "select": [
    {
      "name": "partition_col",
      "alias": ""
    },
    {
      "name": "input",
      "alias": ""
    }
  ]
}
```

### Why are the changes needed?

This can be useful for moving this computation from drivers to executors if the Python is expensive.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

This PR adds SQL query golden file test coverage.

### Was this patch authored or co-authored using generative AI tooling?

No.

dtenedor committed:
comimt

dtenedor committed:
Merge branch 'master' into scalar-function-udtf-return-analyze
sync

dtenedor committed:
commit

commit

commit

commit

commit

commit

commit

commit

commit

commit

commit

commit

commit

commit

dtenedor commented:
cc @allisonwang-db @ueshin 👀 ❓

allisonwang-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/PythonUDF.scala:
```diff
@@ -252,17 +261,24 @@ case class UnresolvedPolymorphicPythonUDTF(
- *                             contains all metadata returned by the Python UDTF 'analyze' method
- *                             including the result schema of the function call as well as optional
- *                             other information
+ * @param partitionByStrings String representations of the partitioning expressions before parsing.
+ * @param orderByStrings String representations of the ordering expressions before parsing.
+ * @param selectedInputStrings String representations of the selected input expressions before
```
why do we need the string representation here?

dtenedor replied:
I reverted this part.

allisonwang-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/plans/logical/pythonLogicalOperators.scala:
```diff
@@ -274,6 +274,24 @@ case class ArrowEvalPythonUDTF(
 }
 
+/**
+ * A logical plan that evaluates the 'analyze' method of a [[PythonUDTF]] on the executors.
```
Maybe this is beyond the scope of this PR, but could we add more comments/examples here on how a polymorphic UDTF can be transformed into a plan with this node where it's `analyze` method can be evaluated on executors?

dtenedor replied:
Sure, I added some more comments here to help explain.

allisonwang-db commented on sql/core/src/main/scala/org/apache/spark/sql/execution/python/AnalyzePythonUDTFOnExecutorsExec.scala:
```diff
@@ -0,0 +1,77 @@
+      context: TaskContext): Iterator[Iterator[InternalRow]] = {
+    val tableArgs = udtf.children.map(_.isInstanceOf[FunctionTableSubqueryArgumentExpression])
+    val parser = CatalystSqlParser
+    val runner = new UserDefinedPythonTableFunctionAnalyzeRunner(
```
This `UserDefinedPythonTableFunctionAnalyzeRunner` extends from the `PythonPlannerRunner`, which is designed to run Python processes on the driver side. Does it work out of the box on executors?  cc @ueshin

dtenedor replied:
As discussed offline, we need to make this a subclass of `BasePythonRunner`. I can look into that.

dtenedor replied:
This is done.

dtenedor committed:
Merge branch 'master' into scalar-function-udtf-return-analyze
sync

dtenedor committed:
commit

commit

commit

commit

commit

commit

commit

commit

commit

commit

commit

commit

commit

commit

dtenedor committed:
commit

dtenedor committed:
reformat python

dtenedor committed:
add more testing

dtenedor committed:
respond to code review comments

dtenedor reviewed (commented):
Thanks @allisonwang-db for your review! I responded to your comments, and also updated this PR to use "worker.py" per the suggestion from @ueshin. Please take another look.

dtenedor committed:
fix linter

dtenedor committed:
fix linter

dtenedor committed:
fix linter

dtenedor committed:
fix linter

dtenedor committed:
add pickled AnalyzeResult

dtenedor committed:
add pickled AnalyzeResult

dtenedor commented:
cc @allisonwang-db @ueshin friendly ping, this is ready for review

dtenedor committed:
fix linter

dtenedor commented:
I am going to close this since I am getting no response on the PR; will reopen later if we want to implement this feature.
